vpr/src/route/route_utils.cpp

```diff
@@ -0,0 +1,528 @@
+#include "draw_debug.h"
+#include "draw_global.h"
+#include "draw_types.h"
+#include "net_delay.h"
+#include "overuse_report.h"
+#include "place_and_route.h"
+#include "route_debug.h"
+
+#include "VprTimingGraphResolver.h"
+#include "tatum/TimingReporter.hpp"
+
+bool check_net_delays(const Netlist<>& net_list, NetPinsMatrix<float>& net_delay) {
+    constexpr float ERROR_TOL = 0.0001;
+
+    unsigned int ipin;
+    auto net_delay_check = make_net_pins_matrix<float>(net_list);
+
+    load_net_delay_from_routing(net_list, net_delay_check);
+
+    for (auto net_id : net_list.nets()) {
+        for (ipin = 1; ipin < net_list.net_pins(net_id).size(); ipin++) {
+            if (net_delay_check[net_id][ipin] == 0.) { /* Should be only GLOBAL nets */
+                if (fabs(net_delay[net_id][ipin]) > ERROR_TOL) {
+                    VPR_ERROR(VPR_ERROR_ROUTE,
+                              "in timing_driven_check_net_delays: net %lu pin %d.\n"
+                              "\tIncremental calc. net_delay is %g, but from scratch net delay is %g.\n",
+                              size_t(net_id), ipin, net_delay[net_id][ipin], net_delay_check[net_id][ipin]);
+                }
+            } else {
+                float error = fabs(1.0 - net_delay[net_id][ipin] / net_delay_check[net_id][ipin]);
+                if (error > ERROR_TOL) {
+                    VPR_ERROR(VPR_ERROR_ROUTE,
+                              "in timing_driven_check_net_delays: net %d pin %lu.\n"
+                              "\tIncremental calc. net_delay is %g, but from scratch net delay is %g.\n",
+                              size_t(net_id), ipin, net_delay[net_id][ipin], net_delay_check[net_id][ipin]);
+                }
+            }
+        }
+    }
+
+    return true;
+}
+
+// In heavily congested designs a static bounding box (BB) can
+// become problematic for routability (it effectively enforces a
+// hard blockage restricting where a net can route).
+//
+// For instance, the router will try to route non-critical connections
+// away from congested regions, but may end up hitting the edge of the
+// bounding box. Limiting how far out-of-the-way it can be routed, and
+// preventing congestion from resolving.
+//
+// To alleviate this, we dynamically expand net bounding boxes if the net's
+// *current* routing uses RR nodes 'close' to the edge of it's bounding box.
+//
+// The result is that connections trying to move out of the way and hitting
+// their BB will have their bounding boxes will expand slowly in that direction.
+// This helps spread out regions of heavy congestion (over several routing
+// iterations).
+//
+// By growing the BBs slowly and only as needed we minimize the size of the BBs.
+// This helps keep the router's graph search fast.
+//
+// Typically, only a small minority of nets (typically > 10%) have their BBs updated
+// each routing iteration.
+size_t dynamic_update_bounding_boxes(const std::vector<ParentNetId>& updated_nets) {
```

Comment:
Make sure there is a reasonable comment in the .h file too; could move this there or do a more brief .h comment.